pkg/controller/che/tls-secrets.go

```diff
@@ -0,0 +1,287 @@
+//
+// Copyright (c) 2020 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+package che
+
+import (
+	"context"
+	"time"
+
+	orgv1 "github.com/eclipse/che-operator/pkg/apis/org/v1"
+	"github.com/eclipse/che-operator/pkg/deploy"
+	"github.com/sirupsen/logrus"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbac "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// shouldReturn specifies whether caller of HandleCheTLSSecrets should return from reconcile loop.
+var shouldReturn = true
+
+// HandleCheTLSSecrets handles TLS secrets required for Che deployment
+func HandleCheTLSSecrets(checluster *orgv1.CheCluster, r *ReconcileChe) (bool, reconcile.Result, error) {
+	shouldReturn = true
+	reconcileResult, err := processCheTLSSecrets(checluster, r)
+	return shouldReturn, reconcileResult, err
+}
+
+func processCheTLSSecrets(checluster *orgv1.CheCluster, r *ReconcileChe) (reconcile.Result, error) {
+	var cheTLSSecretName = checluster.Spec.K8s.TlsSecretName
+	const cheTLSSelfSignedCertificateSecretName string = "self-signed-certificate"
+
+	// ===== Check Che TLS certificate ===== //
+
+	cheTLSSecret := &corev1.Secret{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: checluster.Namespace, Name: cheTLSSecretName}, cheTLSSecret)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			// Error reading secret info
+			logrus.Errorf("Error getting Che secert \"%s\": %v", cheTLSSecretName, err)
+			return reconcile.Result{}, err
+		}
+
+		// Che TLS secret doesn't exist, generate a new one
+
+		// Remove Che CA certificate secret if any
+		cheCASelfSignedCertificateSecret := &corev1.Secret{}
+		err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: checluster.Namespace, Name: cheTLSSelfSignedCertificateSecretName}, cheCASelfSignedCertificateSecret)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				// Error reading secret info
+				logrus.Errorf("Error getting Che secert \"%s\": %v", cheTLSSecretName, err)
+				return reconcile.Result{}, err
+			}
+			// Che CA certificate doesn't exists (that's expected at this point), do nothing
+		} else {
+			// Remove Che CA secret because Che TLS secret is missing (they should be generated together).
+			if err = r.client.Delete(context.TODO(), cheCASelfSignedCertificateSecret); err != nil {
+				logrus.Errorf("Error deleting Che TLS secret: %v", err)
+				return reconcile.Result{}, err
+			}
+		}
+
+		// Prepare permissions for the certificate generation job
```

Comment:
Why do we need extra permissions?

Reply:
We are creating job there and the job should be able to creates secrets. By default it doesn't have permissions to do so and therefore will fail. To handle that we need to create service account, role and role binding to grant permissions to the job to create secrets. Also these permissions cannot be wider than permissions of the unit which gives them.

Reply:
+1 Could we reuse operator ServiceAccount https://github.com/eclipse/che-operator/blob/master/deploy/service_account.yaml?

Reply:
I think it's possible, but the service account you suggest to use has much wider permissions than the job need. I tried to stick to minimal required permissions for the job.

Reply:
Ok, minimal permission per job is a good point.